prog/hints.go

```diff
@@ -83,10 +123,182 @@ func generateHints(p *Prog, compMap CompMap, c *Call, arg Arg, exec func(newP *P
+				newV = transform(newV)
+				replacersSet[newV] = true
+			}
+		}
+	}
+	// Each transform is related to a different way of encoding an int.
+	transforms := []func(uint64) uint64{
+		// Identity transform.
+		func(v uint64) uint64 { return v },
+		// Revert order of bytes (Big Endian <-> Little Endian) and leave
+		// trailing zero bytes.
+		func(v uint64) uint64 { return reverse(v, false) },
+		// Revert order of bytes (Big Endian <-> Little Endian) and don't leave
+		// trailing zero bytes.
+		func(v uint64) uint64 { return reverse(v, true) },
+	}
+	for _, t := range transforms {
+		f(t)
+	}
+
+	return replacersSet
+}
+
+// Returns an array of different mutations of an ConstArg value. Each of these
+// mutations should be matched against comparison operands.
+func getMutationsForConstVal(v uint64) []uint64 {
+	values := []uint64{v}
+	mutations := []func(uint64) []uint64{
+		shrinkMutation,
+		expandMutation,
+	}
+	for _, m := range mutations {
+		values = append(values, m(v)...)
+	}
+	return values
+}
+
+// Mutation 1: shrink values. Useful in cases like:
+//	void f(int64 v64) {
+//		v32 = (int32) v32;
+//		if (v32 == -1) {...};
+//	}
+// If v64 < 0 and v64's value fits into 32 bits (e.g. v64 = -2), then:
+// uint64(v64) = 0xfffffffffffffffe and uint64(v32) = 0xfffffffe.
+// Thus, the comparison will be (0xfffffffe vs 0xffffffff), and we'll be
+// unable to find the operand in the input stream.
+// This is why we need to check for the 0xff... prefix.
+// If v64 >= 0, we want to match comparisons of type int8(v64) == 0x42,
+// thus we need to drop 7 bytes of v64 (they might be filled with random trash).
+// A solution for both cases is to just drop the higher bytes.
+func shrinkMutation(v uint64) (values []uint64) {
+	for _, half := range rightHalves {
+		values = append(values, v&half)
+	}
+	return
+}
+
+// Mutation 2: expand values. Useful in cases like:
+//	void f(int32 v32) {
+//		v64 = (int64) v32;
+//		if (v32 == -1) {...};
+//	}
+// Same logic as for shrink mutation applies.
+// An example: for 0xab we want to obtain the following values:
+// 0xffab, 0xffffffab, 0xffffffffffffffab (these are the values we get
+// if we cast an int8 c = 0xab to int16, int32, int64).
+func expandMutation(v uint64) (values []uint64) {
+	if v == 0 {
+		return
+	}
+	// Find the most significant byte and its index, to check if the value < 0.
+	msByteValue, msByteIndex, _ := getMostSignificantByte(v)
+	if !valueIsNegative(msByteValue, msByteIndex) {
+		return
+	}
+	for _, size := range []int{2, 4, 8} {
+		if size > msByteIndex+1 {
+			v = v | leftHalves[size]
+			values = append(values, v)
+		}
+	}
+	return
+}
+
+// Reverses the byte order of v.
+// If cutBytes == true, then cuts the trailing bytes, e.g.:
+// reverse(0xefbeadde, true) = 0xdeadbeef
+// reverse(0xefbeadde, false) = 0xdeadbeef00000000
+func reverse(v uint64, cutBytes bool) uint64 {
+	_, msByteIndex, bytes := getMostSignificantByte(v)
```

Comment:
This explodes my brain. Please either add tests that explain what scenarios we want to handle with this, or remove reverse for now. Things I can't understand:
 - why are looking at msb here, but ignore it in shrink
 - if it is useful to apply this to all results of shrink/expand or not
 - if we need to do some big-endian sign extension for values or not
 - if we need to shift by other values or not (e.g. convert 0xabcd to big-endian and put into a 4 byte value, currently we only try to put it into a 2-byte and 8-byte values)